proto/provenance/metadata/v1/scope.proto

```diff
@@ -0,0 +1,234 @@
+syntax = "proto3";
+package provenance.metadata.v1;
+
+option go_package = "github.com/provenance-io/provenance/x/metadata/types";
+
+option java_package        = "io.provenance.metadata.v1";
+option java_multiple_files = true;
+
+import "gogoproto/gogo.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/descriptor.proto";
+import "provenance/metadata/v1/specification.proto";
+
+/**
+
+Overview
+
+The structures in this file are used to efficiently store the data for the metadata module.  The structure the core
+concept of the  Metadata module is a Scope that contains a list of Records/Facts.  Updates to scopes are performed
+in a session with details persisted in a RecordGroup.  Each record group contains a specification that details
+constraints on parties that must sign, inputs, and outputs that can be recorded.  Each Scope may contain a list of
+allowed specifications that may be used.
+
+
+Data Structures
+
+  Metadata Scope
+
+  A metadata scope contains a collection of records of data, each made up of a hash, data type, and ownership
+  information that establishes provenance.
+
+  Record Groups
+
+  Typically multiple records will be established within a scope at the same time (see: Contract Memorialization).  A
+  Record Group contains context for the recording of these records linking a set of records into a common process/
+  execution associated with a specificaiton indicating allowed and required values.
+
+  Record
+
+  A single record collects information about a process that generated it, a collection of inputs, and one or more
+  outputs into a structure identified by a name that is unique within a scope.
+
+
+KVStore State Data
+
+Iterators
+- Get All Scopes (<scope prefix>) []Scope
+- Get All RecordGroups (<group prefix>:<scope_id>) []RecordGroup
+  * This method would be used to list all current session/transaction contexts with Records set on the scope
+- Get All Scope Records (<records prefix>:<scope_id>) : []Record
+
+
+Direct Access
+- Get Scope (<scope prefix>:<scope_id>) Scope
+- Get RecordGroup (<group prefix>:<scope_id>:<group_id>) : RecordGroup
+- Get Record (<record prefix>:<scope_id>:<record_name_hash>) : Record
+
+
+Indexes (special kvstore records for efficient cross reference/queries)
+- Get RecordGroup Records - use filtered iterator over all scope records
+- Get all scopes for a party/address?
+- Get all scopes of a specific type (scope spec)?
+- Get all record groups of a specific specification type?
+- Get all Records of a type?
+
+
+*/
+
+// Scope defines a root reference for a collection of records owned by one or more parties.
+message Scope {
+  option (gogoproto.goproto_stringer) = false;
+
+  // Unique ID for this scope.  Implements sdk.Address interface for use where addresses are required in Cosmos
+  bytes scope_id = 1 [
```

Comment:
cool, but won't this be a huge change for p8e?  Are they aware that scope ID will be changing from UUID to address?

Reply:
The response types for messages in queries will need to return the UUID.  Same for appropriate events.  The address class has some extensive methods/tests/mappings to make that easy to accomplish.